src/library_async.js

```diff
@@ -749,10 +806,14 @@ mergeInto(LibraryManager.library, {
         runAndAbortIfError(Module['_asyncify_stop_rewind']);
         Asyncify.freeData(Asyncify.currData);
```

Comment:
this previously-existing line frees the context when we finish a rewind. is that not a problem for this patch?

Reply:
It's not a problem, at least with the intended pattern of usage in mind. I can see it being a problem if you try to swap into the same fiber more than once, without having updated it (by swapping *out* of it first). But that can only happen if something like this is going on (pseudo-code):

```c
// Start at FiberA

FiberA: {
/* 1 */  emscripten_fiber_swap(FiberA, FiberB);   // goto FiberB line 1; FiberA is updated
/* 2 */  do_stuff();
/* 3 */  emscripten_fiber_swap(FiberA, FiberB);   // UB; FiberB has a dangling asyncify data pointer
}

FiberB: {
/* 1 */  thisFiber = emscripten_fiber_create_from_current_context();  // weird contrived usage
/* 2 */  emscripten_fiber_swap(thisFiber, FiberA);  // goto FiberA line 2; FiberB is NOT updated
/* 3 */  do_stuff();
}
```

I could either try to work around this to allow that kind of usage (if it's good for anything), or just warn about this quirk in the documentation.

Reply:
I see, thanks. Maybe we could assert that `emscripten_fiber_create_from_current_context` is only called from the main context (not in a fiber) somehow? (If it's not easy, at least documenting this possible issue would be good, on `emscripten_fiber_create_from_current_context` entry in the docs.)

Otherwise this looks good! Nice work.

Reply:
The runtime doesn't actually distinguish between the main context and fibers, so that's not completely straightforward. It's doable, but I'd rather not introduce more global state just for this purpose.

There is technically nothing wrong with calling `emscripten_fiber_create_from_current_context` from a fiber. You just have to take care not to swap into the old fiber (that you used to enter this context), if you use the new fiber to leave the context.

*Yes, that's very confusing; it's the best explanation I can manage at 3:30 AM*